components/net/bluetooth_thread.rs

```diff
@@ -0,0 +1,371 @@
+
+        let message = BluetoothObjectMsg::BluetoothDevice {
+            id: device.get_address(),
+            name: device.get_name(),
+            device_class: device.get_class(),
+            vendor_id_source: device.get_vendor_id_source(),
+            vendor_id: device.get_vendor_id(),
+            product_id: device.get_product_id(),
+            product_version: device.get_device_id(),
+            appearance: device.get_appearance(),
+            tx_power: device.get_tx_power() as i8,
+            rssi: device.get_rssi() as i8
+        };
+        sender.send(message).unwrap();
+    }
+
+    pub fn gatt_server_connect(&mut self, device_id: String, sender: IpcSender<BluetoothObjectMsg>) {
+        if !self.adapter.is_initialized() {
+            sender.send(BluetoothObjectMsg::Error { error: String::from("No adapter found") }).unwrap();
+            return;
+        }
+
+        let device = match self.adapter.get_device(device_id, true) {
+            Some(d) => d,
+            None => {
+                sender.send(BluetoothObjectMsg::Error { error: String::from("No device found") }).unwrap();
+                return;
+            }
+        };
+
+        //TODO check result
+        device.connect();
+        println!("Connected: {:?}", device.is_connected());
+        let message = BluetoothObjectMsg::BluetoothServer {
+            connected: true
+        };
+        sender.send(message).unwrap();
+    }
+
+    pub fn gatt_server_disconnect(&mut self, device_id: String, sender: IpcSender<BluetoothObjectMsg>) {
+        if !self.adapter.is_initialized() {
+            sender.send(BluetoothObjectMsg::Error { error: String::from("No adapter found") }).unwrap();
+            return;
+        }
+
+        let device = match self.adapter.get_device(device_id, true) {
+            Some(d) => d,
+            None => {
+                sender.send(BluetoothObjectMsg::Error { error: String::from("No device found") }).unwrap();
+                return;
+            }
+        };
+
+        //TODO check result
+        device.disconnect();
+        println!("Connected: {:?}", device.is_connected());
+        let message = BluetoothObjectMsg::BluetoothServer {
+            connected: false
+        };
+        sender.send(message).unwrap();
+    }
+
+    pub fn get_primary_service(&mut self, device_id: String, sender: IpcSender<BluetoothObjectMsg>) {
+        if !self.adapter.is_initialized() {
+            sender.send(BluetoothObjectMsg::Error { error: String::from("No adapter found") }).unwrap();
+            return;
+        }
+
+        let device = match self.adapter.get_device(device_id.clone(), true) {
+            Some(d) => d,
+            None => {
+                sender.send(BluetoothObjectMsg::Error { error: String::from("No device found") }).unwrap();
+                return;
+            }
+        };
+
+        let services = device.get_gatt_services();
+        println!("{:?} service(s) found", services.len());
```

Comment:
Maybe this line should be written in the "else" branch of the "if" below. (if services = 0, it would print it twice)